File: 2021/python/02.py
# flake8: noqa

from python import *

aoc = AdventOfCode("2021", "02", "Dive!", new)


@aoc.part(1)
def part1(items):
    pos = [0, 0]
    for item in items:
        [cmd, val] = item.split(" ")
        if cmd == "forward":
            pos[0] += int(val)
        elif cmd == "up":
            pos[1] -= int(val)
        elif cmd == "down":
            pos[1] += int(val)
    return pos[1] * pos[0]


@aoc.part(2)
def part2(items):
    pos = [0, 0]
    aim = 0
    for item in items:
        [cmd, val] = item.split(" ")
        if cmd == "forward":
            pos[0] += int(val)
            pos[1] += aim * int(val)
        elif cmd == "up":
            aim -= int(val)
        elif cmd == "down":
            aim += int(val)
    return pos[1] * pos[0]


if __name__ == "__main__":
    aoc.solve()

File: 2021/ruby/02.rb
lines = File.readlines("../input/02").map { |line| a, b = line.split; [a, b.to_i] }
# part 1
x = 0
depth = 0

lines.each do |op, val|
    case op
    when 'forward'; x += val;
    when 'down'; depth += val;
    when 'up'; depth -= val;
    end
end
puts x * depth

# part 2
x = 0
depth = 0
aim = 0

lines.each do |op, val|
    case op
    when 'forward'; x += val; depth += aim * val
    when 'down'; aim += val
    when 'up'; aim -= val
    end
end
puts x * depth